i3pystatus support

tobes opened the pull request:
So after the fun of #687 I had a go at i3pystatus modules support.

In some ways they are easier, there is a lot less code but they are also more problematic.  ~~Some modules eg weather cannot easily be added to the py3status config (or at least I haven't worked out an easy way to do it).  Realistically full support will be difficult.~~  They also clash really badly on `format`

Their current release on pypi is partially broken so some modules do not work eg gpu_mem and they have no support for python2.

It's interesting to see what the opposition are doing eg they support double clicks (which is quite easy but gets confusing with the scroll-wheel (is that multi-clicking or just lots of really fast single ones).  Also from playing on i3blocks it seems there are many more mouse buttons available (seven on this laptop) 

I'm not sure if this PR has much value apart from as an experiment but it's here so people can have a play if they wish.

### <a name="i3pystatus_block"></a>i3pystatus_block

Support i3pystatus modules in py3status.

[i3pystatus](https://github.com/enkore/i3pystatus) is an alternative to
py3status and provides a variety of modules.  This py3status module allows
these modules to run and be display inside py3status.

The modules parameters are provide as such
```
i3pystatus_block clock {
    module = 'clock'
    format = [ ('%a %b %-d %b %X', 'America/New_York'), ('%X', 'Etc/GMT+9') ]
}
```

if backend(s) are provided they should be given as a dict with the key being
the backend name and the value being a dict of the backend settings

```
i3pystatus_block weather {
    module = 'weather'
    format = '{condition} {current_temp}{temp_unit}[ {icon}]'
    format += '[ Hi: {high_temp}][ Lo: {low_temp}][ {update_error}]'
    backend = {
        'weathercom.Weathercom': {
            'location_code': '94107:4:US',
            'units': 'imperial',
        }
    }
}
```

Configuration parameters:
  - `module` name of the i3pystatus module *(default None)*

    (additional parameters as per the i3pystatus module)

Requires:
  - `i3pystatus`

lasers commented:
Here's the list of oppositions. https://wiki.archlinux.org/index.php/I3#i3status
Take them out? For ice cream or something else? Yes I understand... For fun too.

Not possible to link it directly? Like this?
```
i3py3status "weekcal"{
    command="~/src/i3pystatus/i3pystatus/modules/weekcal.py"
}
```
Anyway, I go ahead and use `aur/i3py3status-git` instead.

My first issue was using i3py3status_block, not i3pystatus_block.
Oh boy. I got confused again. Sorry. 😕 

Is i3py3status better? More modules... Hmm... ❓ 

Hi. Sharing my findings....

```
i3pystatus_block clock{
    module = 'clock'
}
i3pystatus_block cmus{
    module = 'cmus'
}
order += "i3pystatus_block clock"
# order += "i3pystatus_block cmus"
```
clock ── no issue. Mon Jan 23 10:49:05 PM
cmus ── no issue.
clock, cmus ── yes issue! (cmus) raised an exception (TypoError).

I stop now because very problematic if can't use more than one module.

tobes commented:
> Not possible to link it directly? Like this? `command="~/src/i3pystatus/i3pystatus/modules/weekcal.py"`

this has issues about the class name etc but I wanted to keep things simple eg `module = clock`, also uses i3pystatus to do the work.

> My first issue was using i3py3status_block, not i3pystatus_block.

really I'd like to have just used `i3pystatus` but this is not possible. why would you want it to be i3py3status?

> clock, cmus ── yes issue! (cmus) raised an exception (TypoError).

If you get errors sharing is always helpful

> I stop now because very problematic if can't use more than one module.

I have done work especially so that this is possible (eg clock, clock, backlight).  Without the error I can do nothing

lasers commented:
>really I'd like to have just used i3pystatus but this is not possible. why would you want it to be i3py3status?

I don't. It's just a typo that occurred when juggling between both.

>If you get errors sharing is always helpful

Sorry. I assume something simple as this could be reproduced since it's a fresh PR. Here's the error when using only clock, cmus.

```
2017-01-24 04:18:29 INFO Traceback
TypeError: 'str' object does not support item assignment
  File "/home/chris/src/py3status/py3status/module.py", line 99, in prepare_module
    self.module_class.post_config_hook()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 167, in post_config_hook
    self.__class__.__status__ = Status()
  File "/usr/lib/python3.6/site-packages/i3pystatus/core/__init__.py", line 86, in __init__
    self.io = io.StandaloneIO(self.click_events, self.modules, interval)
  File "/usr/lib/python3.6/site-packages/i3pystatus/core/io.py", line 70, in __init__
    self.proto[0]['click_events'] = click_events

2017-01-24 04:18:29 ERROR py3status: post_config_hook in `i3pystatus_block cmus` raised an exception (TypeError) i3pystatus_block.py line 167. Please try to fix this and reload i3wm (Mod+Shift+R)
```
>Without the error I can do nothing

Yes sir. See above.

tobes commented:
> I assume something simple as this could be reproduced since it's a fresh PR. 

a) I don't have cmus installed

b) this looks like a python 3.6 issue (I know I blame everything on 3.6)  I've tried another way to prevent `i3pystatus.Status` being invoked twice - another forced push

lasers commented:
I can do `clock, clock`. I can't do `clock, cpu_freq`. `cpu_freq` works fine alone. 
```
2017-01-24 04:51:08 WARNING post_config_hook in `i3pystatus_block cpu_freq` raised an exception (TypeError) i3pystatus_block.py line 167.
2017-01-24 04:51:08 INFO Traceback
TypeError: 'str' object does not support item assignment
  File "/home/chris/src/py3status/py3status/module.py", line 99, in prepare_module
    self.module_class.post_config_hook()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 167, in post_config_hook
    self.__class__.__status__ = Status()
  File "/usr/lib/python3.6/site-packages/i3pystatus/core/__init__.py", line 86, in __init__
    self.io = io.StandaloneIO(self.click_events, self.modules, interval)
  File "/usr/lib/python3.6/site-packages/i3pystatus/core/io.py", line 70, in __init__
    self.proto[0]['click_events'] = click_events
```
Same error as above... Nothing to do with `cmus`. New git alias works okay.

tobes commented:
> New git alias works okay.

not if that is the error you just got `self.__class__.__status__ = Status()` this line is no longer there

lasers commented:
I saw (forced updates) somewhere. I thought that was it. 😕

New commit. Good! I got `clock`, `cmus`, and `cpu_freq` all running together. Ty for this fix.
I will test nearly all `i3pystatus` modules when I can. You will get your report. Ty again. 👍

lasers commented:
abc_radio ── no issue. requies deps. very cool.
alsa ── no issue. requires deps. white/darkgrey color for on/off.
anybar ── no issue. nice simple feature. https://github.com/tonsky/AnyBar
backlight ── no issue. requires xbacklight.
battery ── no issue.
bitcoin ── no issue. ฿ 908.52
clock ── no issue. Mon Jan 23 10:49:05 PM
cmus ── no issue. 0:03/2:16 (song name here)
cpu_freq ── no issue. 2.34 (in ghz or so)
cpu_usage ── no issue. 05%
cpu_usage_bar ── no issue. `▊▊         `, `▊▊▊▊▊   `, etc.
cpu_usage_graph ── no issue. `▁▁▁_▁▁__▁▁_____` `▁▁▁▁▂▂▅▇▇▇▇▇▇▇▇`, etc.
disk ── no issue. path required in config.
dota2wins ── i skip this. requires dota2py. i no play dota.
dpms ── no issue, but i think onclick might be broken between py3status and i3pystatus.
dpms ── stuck at 'DPMS: off' with white/darkgrey for on/off. no 'DPMS: on' ??
external_ip ── no issue. requires python-geoip dep. annoying.
file ── i don't understand this, can't get this to work. need help. this is probably fine.
github ── no issue. be patient if no status. green 'Github'.
google_calendar ── i skip this. requires several dependencies. probably okay.
gpu_mem ── no issue. yellow '126.0 MiB'
gpu_temp ── no issue. normal '61 °C'

At this point, I think everything will works okay.

iinet ── keyerror exception. i think need to input username/pwd first. skip.
k eyboard_locks ── no issue. `___ ___ ___` --> `CAP ___ ___`
lastfm ── skip. need to configure lastfm. configmissingerror exception.
load ── no issue. `1.60 1.64`
makewatch ── no issue. green `make: idle` -- i have all modules above running since...
makewatch ── see above. both red `make: sleeping` and/or red 'make:running' -- active.
mem ── no issue. yellow `2641.3 MiB`
mem_bar ── no issue. yellow `████    `
moc ── no issue. Same as cmus above.
modsde ── configmissingerror. need user/pwd for that forum. skip.
moon ── no issue. '11.114103829214855 WanCres' mah moon phrase!
mpd ── no issue. clickable mpd pause/play (idk if we got that too)

HALFWAY THROUGH A-Z. BREAK TIME.

lasers commented:
net_speed ── no issue. requires speednet-cli, `↓16.8Mbps ↑5.5Mbps (iTV-3, Inc.)`
network ── no issue. red  `eth0: DOWN`, green `eno1 _______________0.0KB/s`
──────── this could be better by automatically detect interface.
now_playing ── typerror exception. tuple. tuple. tuple.
```
2017-01-24 06:48:15 WARNING Instance `i3pystatus_block now_playing`, user method `run` failed (TypeError) i3pystatus_block.py line 196.
2017-01-24 06:48:15 INFO Traceback
TypeError: 'NoneType' object is not subscriptable
  File "/home/chris/src/py3status/py3status/module.py", line 584, in run
    response = method()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 196, in run
    if isinstance(output['full_text'], tuple):
```
online ── no iusse. just normal `online` or red `offline`
openfiles ── no issue. `open/max: 5565/806992`
openstack_vms ── i skip this. not for me. requires dependency.

I now got an exception from mpd by just leaving it running.
Will comment out all running modules above. top bar overflow.
```
2017-01-24 06:52:05 WARNING Instance `i3pystatus_block mpd`, user method `run` failed (AttributeError) i3pystatus_block.py line 193.
2017-01-24 06:52:05 INFO Traceback
AttributeError: 'NoneType' object has no attribute 'get'
  File "/home/chris/src/py3status/py3status/module.py", line 584, in run
    response = method()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 193, in run
    self.module.run()
  File "/usr/lib/python3.6/site-packages/i3pystatus/mpd.py", line 148, in run
    "title": currentsong.get("Title", ""),
```

openvpn ── i skip this. exception. may have to be configured first.
parcel ── i skip this. don't have a parcel to track. cool tho.
pianobar ── i skip this. install pianobar, configure fifo and event_command, etc.
ping ── no issue. normal '30.3 ms'
plexstatus ── i skip this. need user/passwd,, prot, etc. too much shit to configure.
pomodoro ── does not work out of the box. had to configure something in config first. idk. i failed.
pyload ── eh, i skip this. need to configure user/passwd/keyring, etc.
reddit ── eh, i skip this. need to configure user/passwd/keyring, etc.
redshift ── maybe issue. normal 'On OK'... then red 'redshift exited unexpectedly.'
──────── idk. i have redshift running and it does not pick it up. w/e. my problem.
regex ── idk. i skip this.
runwatch ── no issue displaying red 'firefox', but idk how to use this correctly.
scratchpad ── errors.

```
2017-01-24 07:23:10 WARNING Instance `i3pystatus_block scratchpad`, user method `run` failed (TypeError) i3pystatus_block.py line 196.
2017-01-24 07:23:10 INFO Traceback
TypeError: 'NoneType' object is not subscriptable
  File "/home/chris/src/py3status/py3status/module.py", line 584, in run
    response = method()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 196, in run
    if isinstance(output['full_text'], tuple):
```

ALMOST DONE. BREAK TIME.

tobes commented:
> now_playing ── typerror exception. tuple. tuple. tuple.
> scratchpad ── errors. tuple. tuple. tuple.

These should now be fixed

> I now got an exception from mpd by just leaving it running.

that looks like a bug in i3pystatus

I've added backend support (updated docstring - also added it to first comment here) so modules like weather can now run

```
i3pystatus_block weather{
    module = 'weather'
    format = '{condition} {current_temp}{temp_unit}[ {icon}]'
    format += '[ Hi: {high_temp}][ Lo: {low_temp}][ {update_error}]'
    backend = {
        'weathercom.Weathercom': {
            'location_code': '94107:4:US',
            'units': 'imperial',
        }
    }
}
```

tobes commented:
> reddit ── eh, i skip this. need to configure user/passwd/keyring, etc.

It would be interesting to know if the keyring feature works.

> redshift ── maybe issue. normal 'On OK'... then red 'redshift exited unexpectedly.'
──────── idk. i have redshift running and it does not pick it up. w/e. my problem.

Because of how redshift works you cannot control a running version so the module needs to run it itself.  So you'd have to kill your running version.  It is a pain.  I keep meaning to create a py3status module for this as it is a nice thing to have running.

lasers commented:
sge ── requires lxml. something about cluster, sun grid engine. idk. i skip this.
shell ── does not work with 'xsel', but okay with 'xsel -o'. we just use 'xsel -o', ok?
solaar ── i skip this. i don't want to mess w/ my logitech unifying receiver. there be hardware bug too... where you connect unifying after perfect 54~ times, then the tiny receiver usb died... bad capitactor or something.
spotify - skip. it's now_playing module, but with 'spotify' in `player_name`. smartttt.
swap ── no issue. green `7764.7 MiB`
syncthing ── skip. user/password, dependency. i no use syncthing.
taskwarrior ── no issue. cyan `Task: Pick up a prescription`
temp - no issue. normal `39.0 °C`
text ── no issue. Display static, colored text. COLOR.... pink "Hi tobes!"
timer ── no issue. `T` --> `-04:57` can increase every 5min w/ clicks.
timewarrior ── error -15.
```
2017-01-24 07:44:51 WARNING Instance `i3pystatus_block timewarrior`, user method `run` failed (CalledProcessError) i3pystatus_block.py line 193.
2017-01-24 07:44:51 INFO Traceback
CalledProcessError: Command '['timew', 'export']' died with <Signals.SIGTERM: 15>.
  File "/home/chris/src/py3status/py3status/module.py", line 584, in run
    response = method()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 193, in run
    self.module.run()
  File "/usr/lib/python3.6/site-packages/i3pystatus/timewarrior.py", line 57, in run
    self.loadTrack()
  File "/usr/lib/python3.6/site-packages/i3pystatus/timewarrior.py", line 40, in loadTrack
    tracks_json = subprocess.check_output(['timew', 'export'])
  File "/usr/lib/python3.6/subprocess.py", line 336, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.6/subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
```

uname ── no issue. `Linux 4.8.13-1-ARCH`
uptime ── no issue. `up 178:26`
vk ──  i skip this. vk social network. need to configure user/passwd/keyring, etc.
weekcal ── no issue. nice. `Tue  23[24]25 26 27 28 29  Jan 2017`
whosonlocation ──  i skip this. email/password/session. i don't have them.
window_title ── loaded. no problem. just can't get it to display anything. hmm.
xkblayout ── requires pypi xkbgroup, so i skip this. should be no problem.
zabbix ──  i skip this. no server/user/pwd.



Things worth mentioning. You'll get an ModuleNotFoundError / Exception exception if you
don't have necessary dependencies to run the said module.

Things worth mentioning. You'll get an ConfigMissingError exception if you
don't have necessary option in the said module section.

Remember `string_`, `format_`, `msg_` etc? Well, here, they don't use prefix. See `keyboard_locks` for an example. They type out what it is, done deal.
Lot of good statuses tbh. Hmm. Maybe I see you later. ;-)

lasers commented:
>>    now_playing ── typerror exception. tuple. tuple. tuple.
>>    scratchpad ── errors. tuple. tuple. tuple.
>
>These should now be fixed

Confirmed. No exception/crash.

`scratchpad` -- Got couple of windows in there and that module still does not display anything. still an issue. idk whose issue... py3status or i3pystatus.

`now_playing` works okay. no issues.

tobes commented:
> dpms ── no issue, but i think onclick might be broken between py3status and i3pystatus.
─────── stuck at 'DPMS: off' with white/darkgrey for on/off. no 'DPMS: on' ??

This is their bug `xset -q | grep -q 'DPMS is Enabled'` should be `xset -q | grep 'DPMS is Enabled'`

> timewarrior ── error -15.

At least this is not just a py3status issue :)

> You'll get an ModuleNotFoundError / Exception exception if you
don't have necessary dependencies to run the said module.
You'll get an ConfigMissingError exception if you
don't have necessary option in the said module section.

yeah I should do something better with these

> scratchpad -- Got couple of windows in there and that module still does not display anything. still an issue. idk whose issue... py3status or i3pystatus.

I suspect them but have not looked, but if no crash then I'm happy

Let me know if you get to try some backend modules.

Maybe due to the successes this might be more than just for fun.  I think I will remove the `py3status_format` option as it just makes things a mess.

Thanks for all the testing efforts you have put in here

lasers commented:
No problem. See my redshift fix above. No crash. I'm done testing for now. Cheers. 👍

tobes commented:
> This is their bug xset -q | grep -q 'DPMS is Enabled' should be xset -q | grep 'DPMS is Enabled'

actually it was a different line anyhow I've raised an issue with them https://github.com/enkore/i3pystatus/issues/537

tobes commented:
Updates

* only we do the updates now, before i3pystatus was doing this too

* less work being done so less cpu

* smoother output noticeable on `cpu_usage_graph` etc

* `py3status_format` option removed

imatimba commented:
Hey, thanks for this, very useful.
spotify module works fine.
pulseaudio shows this error on log:
<pre><code>2017-02-07 08:51:05 WARNING post_config_hook in `i3pystatus_block pulseaudio` raised an exception (TypeError) i3pystatus_block.py line 276.
2017-02-07 08:51:05 INFO Traceback
TypeError: '<' not supported between instances of 'int' and 'NoneType'
  File "/usr/lib/python3.6/site-packages/py3status/module.py", line 99, in prepare_module
    self.module_class.post_config_hook()
  File "/home/imatimba/.config/py3status/i3pystatus_block.py", line 276, in post_config_hook
    MIN_CHECK_INTERVAL)</code></pre>
Tried installing it on the usr/lib path for modules just in case but shows the same error.

lasers commented:
Thanks for reporting this. You're not alone.
Same issue confirmed w/ pulseaudio here.
I didn't test those directories.

Another one. `updates/cower` or `updates/pacman` gets you this.
```
2017-02-07 06:53:29 INFO module "i3pystatus_block updates" click_events=new has_kill=False methods=odict_keys(['run'])
2017-02-07 06:53:29 WARNING post_config_hook in `i3pystatus_block updates` raised an exception (AttributeError) i3pystatus_block.py line 228.
2017-02-07 06:53:29 INFO Traceback
AttributeError: 'str' object has no attribute 'items'
  File "/home/chris/src/py3status/py3status/module.py", line 99, in prepare_module
    self.module_class.post_config_hook()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 228, in post_config_hook
    for key, value in backends.items():
```
Using... 
```
order += "i3pystatus_block updates"
i3pystatus_block updates{
    module = 'updates'
    backends = 'pacman'
}
```

tobes commented:
> Would like to omit _block bit to keep things clean... like i3block

So would I but it is pretty impossible because of how the imports work so it can be anything but `i3pystatus` I'm open to suggestions.

@imatimba @lasers thanks for the bug reports I'll look when I have some time

tobes commented:
@imatimba 

I've had a look and fixed the problem you hit.  I've done a forced push so you'll need to make sure you pull the latest changes. you'll need a `git fetch`.  Thanks for the bug report and feedback.


@lasers the config you need is this, I agree it is ugly, but it works for all backends.  I've added another example to the docstring.

```
i3pystatus_block updates {
    module = 'updates'
    backends = {'pacman.Packman': {}}
}
```

imatimba commented:
@tobes  
Now there is no error on log:
```
2017-02-07 20:53:10 INFO loading module "i3pystatus_block pulseaudio" from /home/imatimba/.config/py3status/i3pystatus_block.py
2017-02-07 20:53:10 INFO module "i3pystatus_block pulseaudio" click_events=new has_kill=False methods=odict_keys(['run'])
```
But is not showing on the bar, using defaults
```
i3pystatus_block pulseaudio {
    module = 'pulseaudio'
}
```

@lasers can you confirm?

lasers commented:
Pulseaudio works for me. Scrollable wheel to adjust volume, color toggling on clicks and a minor error notification because I don't have installed `pavucontrol`. It seems okay on my side. Here's the exception log just in case.
```
2017-02-07 19:40:04 INFO trying to dispatch event to module "i3pystatus_block pulseaudio"
2017-02-07 19:40:04 WARNING on_click event in `i3pystatus_block pulseaudio` failed (AttributeError) i3pystatus_block.py line 158.
2017-02-07 19:40:04 INFO Traceback
AttributeError: 'PulseAudio' object has no attribute 'pavucontrol'
  File "/home/chris/src/py3status/py3status/module.py", line 541, in click_event
    click_method(event)
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 315, in on_click
    self._click_timer.event(event['button'])
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 178, in event
    self.trigger()
  File "/home/chris/src/py3status/py3status/modules/i3pystatus_block.py", line 158, in trigger
    callback_function = getattr(self.parent.module, callback_name)
```
i3pystatus_block updates does not work for me... No print.... with...
* `backends = {'pacman.Pacman': {}}`
* `backends = {'cower.Cower': {}}`
* `backends {'pacman.Pacman': {}, 'cower.Cower':{}}`

imatimba commented:
@lasers ok thanks, I'll check my config then

lasers commented:
I used this.
```
i3pystatus_block pulseaudio {
    module = 'pulseaudio'
}
order += "i3pystatus_block pulseaudio"
```

imatimba commented:
So after testing different things, doing a clean git clone of the branch, disabling other modules, etc; I think I found the problem.  
If the clock module is not present and enabled with a custom format like the one in op, pulseaudio doesn't show in the bar. The separator does.  
If I remove the clock format same thing happens. 

Testing with this config:

```
general {
        colors = true
        interval = 5
}

order += "i3pystatus_block spotify"
i3pystatus_block spotify {
    module = 'spotify'
}

order += "i3pystatus_block pulseaudio"
i3pystatus_block pulseaudio {
    module = 'pulseaudio'
}

order += "i3pystatus_block clock"
i3pystatus_block clock {
    module = 'clock'
    format = [ ('%a %-d %b %X', 'America/New_York'), ('%X', 'Etc/GMT+9') ]
}

```

Clock enabled:

> 2017-02-08 01:09:29 INFO method run returned {'color': '#FFFFFF', 'full_text': '♪: 15', 'instance': 'pulseaudio', 'name': 'i3pystatus_block'} 

Clock disabled:

> 2017-02-08 01:11:27 INFO method run returned {'full_text': '', 'instance': 'pulseaudio', 'name': 'i3pystatus_block'}

lasers commented:
I can't replicate this behavior w/ your config above minus spotify. Clock format too.
Clock and pulseaudio run together flawlessly here. :+1: 

I was able to get your result by killing pulseaudio. I think the problem may be related to initializing pulseaudio. Confirm that it's running. If it's running, try restarting i3wm one more time because I wasn't able to display it afterward after running without doing a i3wm restart.

tobes commented:
Ok, The problem was to do with updating.  Thanks for the test config. This is a little tricky as we don't actually know when the module updates.  Also pulseaudio doesn't have a interval param so was being treated as not needing an update.

I now do the updates more regularly.  We now periodically check for a change.  Currently not the best logic (I might change it) but it should now work better for modules.  Updates are checked 1, 2, 4, 8, 16, every 30 seconds after a change to try to catch them.

imatimba commented:
Thanks! It's working fine now.

lasers commented:
Test result. Immediate `Updates: 6`. Yay.
After updating, I wait more than 2min... Still stuck at `Updates: 6`

Running `checkupdates` myself confirmed no updates.

tobes commented:
> After updating, I wait more than 2min... Still stuck at Updates: 6
>
> Running checkupdates myself confirmed no updates.

that's just i3pystatus doing it's thing.  You might end up waiting an extra 30 seconds for the update but it shouldn't be worse than that.  Anyhow good that it is now working.

lasers commented:
15 minutes in and still `Updates: 6`. I'm running a stopwatch. ;-)

tobes commented:
looking at the code

>  "Default interval is set to one hour."

enjoy the show

lasers commented on py3status/modules/i3pystatus.py:
```diff
@@ -0,0 +1,328 @@
+    (additional parameters as per the i3pystatus module)
+
+Requires:
+    i3pystatus:
```
(Suggestion) `i3py3status: i3status replacement written in python`

tobes replied:
done

lasers commented on py3status/modules/i3pystatus.py:
```diff
@@ -0,0 +1,328 @@
+
+[i3pystatus](https://github.com/enkore/i3pystatus) is an alternative to
+py3status and provides a variety of modules.  This py3status module allows
+these modules to run and be display inside py3status.
```
(Suggestion) `be displayed`
(Question)  Do we want `[i3py3status](https://...)` inside modules? First time here(?).
(Example). Maybe we should add `Examples:` before the descriptions too.... and have the descriptions inside the examples? For consistency, I guess... meaning `configuration parameters` should be the first thing to see, not the examples.

tobes replied:
I think having nice links is good - maybe this isn't the best format.  I'll have a think.  There is a precedent `i3block` but anyhow I want to look at the whole module docs issue.

lasers commented on py3status/modules/i3pystatus.py:
```diff
@@ -0,0 +1,328 @@
+py3status and provides a variety of modules.  This py3status module allows
+these modules to run and be display inside py3status.
+
+The modules parameters are provide as such
```
(Suggestion) `provided as such`

tobes replied:
done

lasers commented on py3status/modules/i3pystatus.py:
```diff
@@ -0,0 +1,328 @@
+
+
+class Py3status:
+
```
(Nit) Add missing empty `docstring` cause it's pretty...

tobes replied:
I think this is a waste of bytes :)

ultrabug replied:
still @tobes, please add them for consistency

lasers commented on py3status/modules/i3pystatus.py:
```diff
@@ -0,0 +1,328 @@
+            callbacks.append((click, dbclick))
+
+        self._click_timer = ClickTimer(self, callbacks)
+        self._due_time = 0
```
(Q) Purpose for this config?

tobes replied:
I think this is redundant I need to do more testing of this module

ultrabug commented:
Friendly bump @tobes ? address comments or not ?

tobes commented:
> Friendly bump @tobes ? address comments or not ?

cool I think I missed these comments, I'll review.

ultrabug commented:
Thanks let's get this in!
